### PR TITLE
HOC-191 Implement RFD-1 - Block Registry

### PR DESCRIPTION
**Pull request by gorbak25**

### CODE REVIEW

File: app/agent/block-registry/registry.service.ts [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+      case EXPOSE_METHOD.BLOCK_DEV: {
+        const t1 = performance.now();
+        // Get the tcmuSubtype
+        const tcmuSubtype = await this.getTCMUSubtype();
+        const tcmuStorageObjectId = `${tcmuSubtype}_${what}`;
+        const tcmuPath = path.join(this.paths.tcmuHBA, tcmuStorageObjectId);
+        // Create a storage object on that HBA, if it already exists then check for a LUN
+        let aluaMembers: string[] = [];
+        await fs.mkdir(tcmuPath).catch(async (err: Error) => {
+          if (!err.message.startsWith("EEXIST")) throw err;
```
Comment (hugodutka) [inline app/agent/block-registry/registry.service.ts:734]:
```suggestion
        await fs.mkdir(tcmuPath).catch(async (err) => {
          if (!err?.message?.startsWith("EEXIST")) throw err;
```

File: app/agent/block-registry/registry.service.ts [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+          //          waiting, due to this we request a flock in unblocking mode and sleep for some time.
+          //          TODO: Consider writing a rust addon for flocks cause there is no good library for this in node .-.
+          const enableFd = await fs.open(path.join(tcmuPath, "enable"), "w");
+          let tcmuAttachFailed = false;
+          try {
+            let flockAcquired = false;
+            while (!flockAcquired) {
+              try {
+                await new Promise((resolve, reject) => {
+                  flock(enableFd.fd, "exnb", (err) => {
```
Comment (hugodutka) [inline app/agent/block-registry/registry.service.ts:753]: you can use promisify instead: https://nodejs.org/api/util.html#utilpromisifyoriginal

File: app/agent/block-registry/registry.service.ts [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+                      resolve(void 0);
+                    } else {
+                      reject(err);
+                    }
+                  });
+                });
+              } catch (err) {
+                if (
+                  !(err as Error).message.startsWith("EAGAIN") &&
+                  !(err as Error).message.startsWith("EWOULDBLOCK")
```
Comment (hugodutka) [inline app/agent/block-registry/registry.service.ts:764]:
```suggestion
                  !err?.message?.startsWith("EAGAIN") &&
                  !err?.message?.startsWith("EWOULDBLOCK")
```

File: app/agent/block-registry/registry.service.ts [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+              let lunClaimed = false;
+              while (!lunClaimed) {
+                await fs.mkdir(lunPath).catch(catchAlreadyExists);
+                try {
+                  await fs
+                    .symlink(tcmuPath, path.join(lunPath, tcmuStorageObjectId))
+                    .catch(catchAlreadyExists);
+                  lunClaimed = true;
+                } catch (err) {
+                  if (!(err as Error).message.includes("ENOENT")) throw err;
```
Comment (hugodutka) [inline app/agent/block-registry/registry.service.ts:891]:
```suggestion
                  if (!err?.message?.includes("ENOENT")) throw err;
```

File: rfd/0001-block-registry.md [hunk 5]
```diff
@@ -30,53 +30,70 @@ OverlayBD is an OCI-compatible container format from Alibaba that meets our requ
...
+Hocus agent requires access to `configfs` and `devtmpfs`, both of which are bind mounted into the agent container from the host. `devtmpfs` should be present at `/dev/hocus` and `configfs` at `/sys/kernel/config`. Limiting privilidges using udev is not sensible as the agent container MUST run in priviledged mode in order to create and manage network namespaces for the workspaces and udev introduces additionaly complexity and latency to the system. The block registry MUST consider that multiple agents might be running on the same computer, and it DOES NOT have exclusivity over `/dev/hocus` or `/sys/kernel/config`.
```
Comment (hugodutka) [inline rfd/0001-block-registry.md:36]:
```suggestion
Hocus agent requires access to `configfs` and `devtmpfs`, both of which are bind mounted into the agent container from the host. `devtmpfs` should be present at `/dev/hocus` and `configfs` at `/sys/kernel/config`. Limiting privileges using udev is not sensible as the agent container MUST run in privileged mode in order to create and manage network namespaces for the workspaces and udev introduces additionaly complexity and latency to the system. The block registry MUST consider that multiple agents might be running on the same computer, and it DOES NOT have exclusivity over `/dev/hocus` or `/sys/kernel/config`.
```

**Review by hugodutka — APPROVED** [review]
Great job! I'm impressed by all the effort you put into making this work.

File: app/agent/block-registry/registry.service.test.ts [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+          `Failed run id: ${runId}. Please investigate ../hocus-resources/tests/${runId}`,
+        );
+        keepStorage = true;
+      }
+      throw err;
+    } finally {
+      if (!opts.skipInit) {
+        // First check the kernel logs cause if we borked the kernel then
+        // all sanity is gone and the cleanup will probably hang .-.
+        await ensureKernelDidNotBlewUp();
```
Comment (gorbak25) [inline app/agent/block-registry/registry.service.test.ts:150]:
```suggestion
        await ensureKernelDidNotBlowUp();
```

File: app/agent/block-registry/registry.service.ts [hunk 7]
```diff
@@ -0,0 +1,1048 @@
...
+              .catch(catchAlreadyExists);
+          else {
+            await fs.mkdir(requiredPath).catch(catchAlreadyExists);
+          }
+        }),
+      );
+    }
+    await waitForPromises(tasks);
+
+    // Play some starcraft, i found ZERO info what is the nexus in tcm_loop .-. I have no idea what this does
```
[the remark this answers is not in the document]
Comment (gorbak25) [reply app/agent/block-registry/registry.service.ts:200]: Let me check if it breaks if I remove  that line.  In all usage examples of tcm_loop the nexus was always set, there are some references to that name in the official SCSI spec.

File: app/agent/block-registry/registry.service.ts (hunk 2 above)
Comment (gorbak25) [reply app/agent/block-registry/registry.service.ts:803]: Just tried that and it doesn't work - `promisify` expects an `(err, value)` callback, where the flock library uses an `(err_or_null)` callback. Additionally promisify converted the `(number, string, callback)` function signature into a `(number, numer)` type sigrature.

File: app/agent/block-registry/registry.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (gorbak25) [reply app/agent/block-registry/registry.service.ts:200]: Tested the code without the nexus. There is no error but no HDD appears in the system. https://www.sanfoundry.com/what-is-iscsi-it-nexus/ I still don't understand why is the nexus necessary but it is.

File: app/agent/block-registry/registry.service.ts [hunk 8]
```diff
@@ -0,0 +1,1048 @@
...
+      // RO Layer storage
+      layers: path.join(root, "layers"),
+      // RW Layer storage
+      containers: path.join(root, "containers"),
+      // OCI Image manifest storage
+      images: path.join(root, "images"),
+      // Directory for anything temporary which should be nuked after a restart
+      // This can't be /tmp as I need the directory to be on the same partition as the registry
+      // I'm abusing hardlinks to create synchronization primitives
+      // The entire registry is fully lockless, wait free and all operations on it are idempotent
```
[the remark this answers is not in the document]
Comment (gorbak25) [reply app/agent/block-registry/registry.service.ts]: Fixed the comment

**Pull request merged**